clustering/ejb/WFLY-14359_Create_distributable-ejb_subsystem.adoc

```diff
@@ -0,0 +1,188 @@
+that it depends on.
+
+There are at present four features provided by the EJB3 subsystem which depend on clustering abstractions:
+
+* provision of cache factories, used to store stateful session bean sessions
+* provision of client mappings registries, used to provide EJB client topology updates
+* provision of singleton services used to support singleton message-driven bean delivery groups
+* provision of distributed timer management, used to support distributed EJB timers (future)
+
+The distributable-ejb subsystem will allow defining different backend providers of these abstractions, which may be configured
+independently of the ejb3 subsystem.
+
+Ultimately, the distributable-ejb subsystem will provide providers for all four of the feature areas listed above.
+But this issue will be restricted to the following scope:
+
+* defining the distributable-ejb subsystem
+* defining a cache factory provider for stateful session beans and integrating that provider into the ejb3 subsystem
+* defining a client mappings registry provider for the EJB client and integrating that provider into the ejb3 subsystem
+
+The remaining two features will be completed in separate RFEs:
+
+** WFLY-7628 will address this for distributed timer management
+** WFLY-15432 will address this for singleton service
+
+== Issue Metadata
+
+=== Issue
+
+* https://issues.redhat.com/browse/WFLY-14953
+
+=== Related Issues
+
+* https://issues.redhat.com/browse/WFLY-7628
+
+=== Dev Contacts
+
+* mailto:{email}[{author}]
+
+=== QE Contacts
+
+* mailto:tborgato@redhat.com[Tommaso Borgato]
+
+=== Testing By
+// Put an x in the relevant field to indicate if testing will be done by Engineering or QE.
+// Discuss with QE during the Kickoff state to decide this
+* [x] Engineering
+
+* [] QE
+
+=== Affected Projects or Components
+
+Clustering, EJB
+
+=== Other Interested Projects
+
+=== Relevant Installation Types
+// Remove the x next to the relevant field if the feature in question is not relevant
+// to that kind of WildFly installation
+* [x] Traditional standalone server (unzipped or provisioned by Galleon)
+
+* [x] Managed domain
+
+* [x] OpenShift s2i
+
+* [x] Bootable jar
+
+== Requirements
+
+=== Hard Requirements
+
+The distributable-ejb subsystem:
+
+* Must allow configuration of one or more named bean management providers
+* Must identify a default bean management provider from amongst those defined in the subsystem
+
+The ejb3 subsystem:
+
+* Must allow integration of cache factories with named bean management providers defined in the distributable-ejb subsystem
+* Must cause the default bean management provider to be used when a distributable cache factory does not specify a bean management provider
+
+=== Nice-to-Have Requirements
+
+The ability to allow overriding the default bean manager for a whole deployment is not a hard requirement, but a nice-to-have.
```

Comment:
Yes, the current round of  changes affect all deployments, unless specifically overridden (i.e. we can still use the @Cache annotation to override a cache used for a specific SFSB). However, there is at this stage no mechanism to configure providers on a per-deployment basis.  This is discussed in the Deployments section.

Reply:
> Yes, the current round of changes affect all deployments, unless...

Ok, so the hard requirements warrant for this to happen.

>  However, there is at this stage no mechanism to configure providers on a per-deployment basis. This is discussed in the Deployments section.

I am I thought this _Nice-to-have requirement is actually defining the need for a functionality that would allow to override the configuration at the deployment level. Indeed, the `Deployments` section reads as follows:

```
A new deployment configuration namespace, specified via jboss-all.xml or a separate distributable-ejb.xml, will be introduced to permit specifying default providers for bean management (resp. client-mappings management) on a per-deployment basis
```

therefore I am not sure I understand your statement. Will the requirement be implemented? Will the feature implementation allow for the user to override the configuration per-deployment too?

Reply:
The ability to override bean management providers and client mappings registry providers at the deployment level, via a distributable-ejb.xml file or otherwise, is planned, and so I mentioned it here is the nice-to-have section. However, given that this PR is already very big and is complete as a unit of functionality, adding the deployment-specific configuration (which involves adjusting many deployment unit processors which itself is quite involved) was better left as a separate and subsequent RFE. So to answer your questions: no, the deployment specific feature will be handled in a later RFE and no, the current RFE will not allow for the user to override the distributable-ejb related configuration on a per-deployment basis. If this is unclear, I can try to rewrite it or just remove it altogether.